### PR TITLE
[fx] add SymPy assumptions to `FloorDiv`

### PR DESCRIPTION
**Pull request by nkaretnikov**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #91376
* #93143
* #93142
* __->__ #93185

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: test/test_dynamic_shapes.py [hunk 1]
```diff
@@ -479,9 +480,6 @@ def print_seen():
     ('floordiv', 'SymFloat', 'int'),  # Scalars are not close!
     ('floordiv', 'float', 'SymInt'),  # Scalars are not close!
     ('floordiv', 'SymFloat', 'SymInt'),  # Scalars are not close!
-    ('floordiv', 'SymInt', 'float'),  # Cannot convert complex to float
-    ('floordiv', 'int', 'SymFloat'),  # Cannot convert complex to float
-    ('floordiv', 'SymInt', 'SymFloat'),  # Cannot convert complex to float
```
Comment (nkaretnikov) [inline test/test_dynamic_shapes.py:484]: Fixes an existing test as well

File: test/test_dynamic_shapes.py [hunk 2]
```diff
@@ -611,5 +609,64 @@ def test_method(self, fn, first_type, second_type):
...
+            7 * FloorDiv(6.28, (FloorDiv(6.28, 3.14))),
+        )
+
+        for expr in exprs:
+            self.assertEqual(expr, result)
+            self.assertEqual(expr.doit(deep=False), result)
+            self.assertEqual(expr.doit(deep=True), result)
+            self.assertEqual(sympy.simplify(expr), result)
+            self.assertEqual(shape_env.simplify(expr), result)
+            self.assertEqual(shape_env.evaluate_expr(expr), result)
```
Comment (nkaretnikov) [inline test/test_dynamic_shapes.py:631]: All cases simplify now

File: test/test_dynamic_shapes.py [hunk 3]
```diff
@@ -611,5 +609,64 @@ def test_method(self, fn, first_type, second_type):
...
+                return x.is_integer is False and x.is_real is False and x.is_complex
+
+            # In regular Python, x//x == 1.0 if x is a float, but FloorDiv
+            # always returns an integer 1 when both args are the same object.
+            # This even works for Symbols with no assumptions specified.
+            if base is divisor:
+                self.assertTrue(op.is_integer)
+                self.assertTrue(op.is_real)
+            elif base.is_integer and divisor.is_integer:
+                self.assertTrue(op.is_integer)
```
Comment (nkaretnikov) [inline test/test_dynamic_shapes.py:662]: We now set the integer assumption for integer inputs

File: torch/fx/experimental/symbolic_shapes.py [hunk 4]
```diff
@@ -251,21 +252,44 @@ class FloorDiv(sympy.Function):
...
+        # https://docs.sympy.org/latest/guides/custom-functions.html#best-practices-for-eval
         @classmethod
         def eval(cls, base, divisor):
             if base == 0:
                 return sympy.Integer(0)
-            if divisor == 1:
+            if isinstance(base, sympy.Integer) and divisor == 1:
                 return base
+            if isinstance(base, sympy.Float) and divisor == 1:
+                return sympy.floor(base)
```
Comment (nkaretnikov) [inline torch/fx/experimental/symbolic_shapes.py:288]: I've split out the case that requires floor since inductor has problems with it sometimes

File: torch/fx/experimental/symbolic_shapes.py [hunk 5]
```diff
@@ -251,21 +252,44 @@ class FloorDiv(sympy.Function):
...
             if isinstance(base, sympy.Integer) and isinstance(divisor, sympy.Integer):
                 return base // divisor
+            if isinstance(base, (sympy.Integer, sympy.Float)) and isinstance(divisor, (sympy.Integer, sympy.Float)):
+                return sympy.floor(base / divisor)
```
Comment (nkaretnikov) [inline torch/fx/experimental/symbolic_shapes.py:292]: Likewise, the floor case is separate just to be safe

File: torch/fx/experimental/symbolic_shapes.py [hunk 6]
```diff
@@ -1040,7 +1064,7 @@ def simplify(self, expr: "sympy.Expr") -> "sympy.Expr":
             for atom in expr.atoms(FloorDiv):
                 base, divisor = atom.args
                 if self.replace(base % divisor) in self.divisible:
-                    div_replacements[atom] = base / divisor
+                    div_replacements[atom] = sympy.floor(base / divisor)
```
Comment (nkaretnikov) [inline torch/fx/experimental/symbolic_shapes.py:1067]: This is important to enable us to optimize:

```
Eq(FloorDiv(integer_expr1, sympy.floor(base/divisor), integer_expr2)
```

If we had this instead, like before:

```
Eq(FloorDiv(integer_expr1, base/divisor, integer_expr2)
```

It would return `False` immediately since the result of `FloorDiv` is an integer only when both args are integers (based on the defined assumptions).

Test that can be used to repro:
```
pytest test/test_proxy_tensor.py -k test_make_fx_symbolic_exhaustive_nn_functional_glu_cpu_float32 -v
```

**Comment by nkaretnikov** [comment]
@ezyang Another attempt to get FloorDiv to work. Here, I implement all assumptions in this one PR.

LMK if this is cool. If so, I'll try to benchmark/profile because I suspect this might be slower compared to master

File: torch/fx/experimental/symbolic_shapes.py [hunk 7]
```diff
@@ -251,21 +252,44 @@ class FloorDiv(sympy.Function):
...
+            return fuzzy_and([self.base.is_integer, self.divisor.is_integer])
 
+        # Automatic evaluation.
+        # https://docs.sympy.org/latest/guides/custom-functions.html#best-practices-for-eval
         @classmethod
         def eval(cls, base, divisor):
             if base == 0:
                 return sympy.Integer(0)
-            if divisor == 1:
+            if isinstance(base, sympy.Integer) and divisor == 1:
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:285]: Don't you want to apply this simplification for any integral expression base aka we should query the assumption to see if it's integral instead of an isinstance check which will only let through literals?

File: torch/fx/experimental/symbolic_shapes.py (hunk 6 above)
Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:1067]: The problem looks legitimate but I think a better solution is to only learn about divisibility when yhe arguments are integer (and not try to do anything clever with float div.) would that work?

**Review by ezyang — COMMENTED** [review]
This looks much much better!

File: test/inductor/test_torchinductor.py [hunk 8]
```diff
@@ -535,10 +535,10 @@ def populate(cls):
...
     def test_indexing_simplification(self):
         sizevars = SizeVarAllocator()
-        i0 = sympy.Symbol("i0")
-        i1 = sympy.Symbol("i1")
-        i2 = sympy.Symbol("i2")
-        r3 = sympy.Symbol("r3")
+        i0 = sympy.Symbol("i0", integer=True)
+        i1 = sympy.Symbol("i1", integer=True)
+        i2 = sympy.Symbol("i2", integer=True)
+        r3 = sympy.Symbol("r3", integer=True)
```
Comment (nkaretnikov) [inline test/inductor/test_torchinductor.py:541]: Is this cool? I've changed it after seeing:

```
AssertionError: Object comparison failed: 128*i2 + i1//1 + 64*ModularIndexing(r3, 1, 2) != i1 + 128*i2 + 64*ModularIndexing(r3, 1, 2)
```

Since the only way to simplify on 1 right away is for i1 to be an integer.

Comment (ezyang) [reply test/inductor/test_torchinductor.py:541]: This LGTM!

File: torch/fx/experimental/symbolic_shapes.py [hunk 9]
```diff
@@ -251,21 +252,44 @@ class FloorDiv(sympy.Function):
...
+        # Automatic evaluation.
+        # https://docs.sympy.org/latest/guides/custom-functions.html#best-practices-for-eval
         @classmethod
         def eval(cls, base, divisor):
             if base == 0:
                 return sympy.Integer(0)
-            if divisor == 1:
+            if base.is_integer and divisor == 1:
                 return base
+            if isinstance(base, sympy.Float) and divisor == 1:
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:287]: Don't you want to also use assumptions here?

**Comment by ezyang** [comment]
So I'm just waiting to see what the updated divisibility logic without a floor looks like, and then I will approve

File: torch/fx/experimental/symbolic_shapes.py (hunk 6 above)
Comment (nkaretnikov) [reply torch/fx/experimental/symbolic_shapes.py]: Updated. The reason for not using assumptions there and using isinstance(..., Integer):

```
# works for Integer type
[ins] In [24]:  (sympy.Integer(42) / 2).is_integer
Out[24]: True

# this is FloorDiv, might call floor for complex exprs
[ins] In [25]:  (sympy.Integer(42) // 2).is_integer
Out[25]: True

# this is FloorDiv, might call floor for complex exprs
[ins] In [26]:  (sympy.Symbol("s", integer=True) // 2).is_integer
Out[26]: True

# None, doesn't work for integer Symbols
[ins] In [27]:  (sympy.Symbol("s", integer=True) / 2).is_integer
```

Comment (nkaretnikov) [reply torch/fx/experimental/symbolic_shapes.py]: Hmm, now that I think of it... Since we allow None in FloorDiv assumptions it should just fail to simplify with `/` for Symbols right away and eval later. Let me do more tests.

Comment (nkaretnikov) [reply torch/fx/experimental/symbolic_shapes.py]: Ah no, using assumptions in simplify breaks this:

```
python benchmarks/dynamo/timm_models.py --ci --training --accuracy \ 
  --dynamic-shapes --device cuda --backend aot_eager \
  --total-partitions 2 --partition-id 0 \
  --output log.txt -k eca_halonext26ts

```

```
RuntimeError: Failed running call_function <built-in function matmul>(*(FakeTensor(FakeTensor(..., device='meta',
           size=((s2**2*s0)//(8*(s2//8)**2), (s2//8)**2, (s2//(s2//8))*((s2*((8*s0)//((s2**2*s0)//(8*(s2//8)**2))))//(s2//8)), 16),
           grad_fn=<TransposeBackward0>), cuda:0), FakeTensor(FakeTensor(..., device='meta', size=(8*s0, (s2//8)**2, 16, 144),
           grad_fn=<TransposeBackward0>), cuda:0)), **{}):
expand: attempting to expand a dimension of length 8*s0!
```

This is the same issue that we saw before: the assumptions system is being too eager, which contradicts floordiv assumptions. So let's not do that.

File: torch/fx/experimental/symbolic_shapes.py [hunk 10]
```diff
@@ -1063,8 +1063,10 @@ def simplify(self, expr: "sympy.Expr") -> "sympy.Expr":
             div_replacements = {}
             for atom in expr.atoms(FloorDiv):
                 base, divisor = atom.args
-                if self.replace(base % divisor) in self.divisible:
-                    div_replacements[atom] = sympy.floor(base / divisor)
+                if self.replace(base % divisor) in self.divisible and (
+                    isinstance(base, sympy.Integer) and isinstance(divisor, sympy.Integer)
+                ):
```
Comment (nkaretnikov) [inline torch/fx/experimental/symbolic_shapes.py:1068]: I don't think it matters that I check the types second because self.divisible is a set

File: torch/fx/experimental/symbolic_shapes.py (hunk 6 above)
Comment (nkaretnikov) [reply torch/fx/experimental/symbolic_shapes.py]: ... but checking against Integer doesn't make sense here because we already have an Integer rule in FloorDiv, so it should eval automatically and this branch would never be executed :(

Comment (nkaretnikov) [reply torch/fx/experimental/symbolic_shapes.py]: @ezyang I don't think there's a way to have replacements unless we use `sympy.floor(base / divisor)` in simplify. The concrete example is https://github.com/pytorch/pytorch/pull/90906#issuecomment-1403061281 and `s2 / 8` is what we need to floor to make it an integer for FloorDiv + Eq to work.

Can I change this back to `sympy.floor(base / divisor)`? Or do you see another solution?

**Comment by nkaretnikov** [comment]
Added floor back to simplify. Context: https://github.com/pytorch/pytorch/pull/93185#discussion_r1090187416

File: torch/fx/experimental/symbolic_shapes.py (hunk 6 above)
Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:1067]: woof. OK, I understand why you want the floor, but I'm just very afraid the floor will make the expressions worse. I think we'll need to consult with @Chillee on this.

Comment (nkaretnikov) [reply torch/fx/experimental/symbolic_shapes.py:1067]: We've discussed this offline. Horace suggested to try defining a custom class that would do truediv, but have assumptions necessary to prove that `s2 / 8` is an integer no matter what. And then we can use it here instead of floor. So let me try that.

Comment (nkaretnikov) [reply torch/fx/experimental/symbolic_shapes.py:1067]: @ezyang @Chillee I've attempted the "clean division" class idea. Unfortunately, I think it doesn't work:

1. `_eval_*` assumption handlers seem to be executed after `eval` and _only if_ the latter returns `None`
2. If you return `None` from `eval`, your new class becomes a node in all exprs, so hard to optimize (same issue as `floor`)
3. Additional maintenance burden from having this custom class (while `floor` captures the semantics of the operation perfectly) + there are some issues with handling it in `_maybe_evaluate_static`, I'm not sure why yet.

Now, let's go back to why we want to avoid `floor` here. IIUC, it's only because it's hard to optimize exprs containing it (+ that inductor bug, but that's separate). Is there anything else?

If it's only optimization, I think we can just keep `floor` and define custom rules to optimize it instead. E.g., Horace suggested that we might want to optimize cases like `floor(x / y) * y`. This is doable by overloading `Mul` and having `floor` elimination logic there. NB: `x` must be divisible by `y` for this to work, counterexample: `sympy.floor(sympy.Integer(7) / 3) * 3 != 7`.

TL;DR: I propose we merge this with `floor` because it passes the tests and the code is clean. Then we can discuss separately how to optimize cases we could be optimizing.

Please LMK if I can merge this or if I need to spend more time attempting to make this work without `floor`.

Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:1067]: OK, while Horace is coming up with a new suggestion, here's what I propose: let's keep the floor, conditional on it not regressing any inductor inference models in CI (we recently added these to CI, so rebase and turn on ciflow/inductor). Assuming no regressions, let's just ship it.

**Review by ezyang — APPROVED** [review]

File: torch/fx/experimental/symbolic_shapes.py (hunk 6 above)
Comment (Chillee) [reply torch/fx/experimental/symbolic_shapes.py:1067]: > If you return None from eval, your new class becomes a node in all exprs, so hard to optimize

We can just define new simplification rules right? I don't think it's that difficult.

> If it's only optimization

IMO, my main objection is that philosophically, `floor(s0/4)` just seems ... wrong, if we know that s0 is divided cleanly by 4. It makes more sense to me to just have a `CleanDiv` object.

> Assuming no regressions, let's just ship it.

If there are no regressions, then I'm fine with it. But I somewhat suspect there will be regressions.

**Comment by ezyang** [comment]
plz rebase

**Comment by nkaretnikov** [comment]
The failures are legit. This breaks because of the floor in simplify (passes without it). I'll try to find a diff solution (that would work for models as well):
```
test/inductor/test_torchinductor.py::CpuTests::test_output_strides_dynamic_shapes_cpu 
```

File: torch/fx/experimental/symbolic_shapes.py [hunk 11]
```diff
@@ -341,7 +341,11 @@ def eval(cls, *args):
 @lru_cache(256)
 def safe_expand(r):
     if hasattr(r, 'expand'):
-        return sympy.expand(r)
+        try:
+            return sympy.expand(r)
+        except RecursionError:
+            log.warning(f"RecursionError in sympy.expand({r})")
+            return r
```
Comment (nkaretnikov) [inline torch/fx/experimental/symbolic_shapes.py:348]: Fix for `test/inductor/test_torchinductor.py::CpuTests::test_output_strides_dynamic_shapes_cpu`. We're doing this for sympy.solve as well, so seems fine, right?

**Comment by ezyang** [comment]
ci merge! such wow!

**Comment by nkaretnikov** [comment]
@ezyang Should we reland this after https://github.com/pytorch/pytorch/pull/93903 lands? I cannot repro https://github.com/pytorch/pytorch/issues/93885 after that

**Comment by ezyang** [comment]
Yes we can try
